trustchain-ion/src/verifier.rs

```diff
@@ -1,26 +1,91 @@
-    MONGO_COLLECTION_OPERATIONS, MONGO_CONNECTION_STRING, MONGO_CREATE_OPERATION,
-    MONGO_DATABASE_ION_TESTNET_CORE, MONGO_FILTER_DID_SUFFIX, MONGO_FILTER_TYPE,
+    PROVISIONAL_INDEX_FILE_URI_KEY, URL,
 };
+use bitcoin::blockdata::transaction::Transaction;
+use bitcoin::hash_types::BlockHash;
 use bitcoincore_rpc::RpcApi;
+use did_ion::sidetree::Delta;
 use futures::executor::block_on;
-use mongodb::{bson::doc, options::ClientOptions, Client};
-use ssi::did_resolve::DIDResolver;
-use std::convert::TryFrom;
-use trustchain_core::resolver::Resolver;
+use futures::TryFutureExt;
+use ipfs_api_backend_actix::IpfsClient;
+use mongodb::bson::doc;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use ssi::did::Document;
+use ssi::did_resolve::{DIDResolver, DocumentMetadata};
+use std::collections::HashMap;
+use std::convert::TryInto;
+use std::str::FromStr;
+use trustchain_core::commitment::{CommitmentError, DIDCommitment};
+use trustchain_core::resolver::{Resolver, ResolverError};
 use trustchain_core::utils::get_did_suffix;
-use trustchain_core::verifier::{Verifier, VerifierError};
+use trustchain_core::verifier::{Timestamp, Verifier, VerifierError};
+
+/// Locator for a transaction on the PoW ledger, given by the pair:
+/// (block_hash, tx_index_within_block).
+type TransactionLocator = (BlockHash, u32);
 
-/// A transaction on the PoW ledger.
-type TransactionIndex = (u32, u32);
+/// Data bundle for DID timestamp verification.
+#[derive(Serialize, Deserialize)]
+pub struct VerificationBundle {
+    /// DID Document.
+    did_doc: Document,
+    /// DID Document Metadata.
+    did_doc_meta: DocumentMetadata,
+    /// ION chunk file.
+    chunk_file: Vec<u8>,
+    /// ION provisional index file.
+    provisional_index_file: Vec<u8>,
+    /// ION core index file.
+    core_index_file: Vec<u8>,
+    /// Bitcoin Transaction (the one that anchors the DID operation in the blockchain).
+    transaction: Vec<u8>,
+    /// MerkleBlock (containing a PartialMerkleTree and the BlockHeader).
+    merkle_block: Vec<u8>,
+    /// Bitcoin block header.
+    block_header: Vec<u8>,
+}
 
-/// Trustchain Verifier for ION DID method. The generic type parameterises the wrapped DID resolver.
+impl VerificationBundle {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        did_doc: Document,
+        did_doc_meta: DocumentMetadata,
+        chunk_file: Vec<u8>,
+        provisional_index_file: Vec<u8>,
+        core_index_file: Vec<u8>,
+        transaction: Vec<u8>,
+        merkle_block: Vec<u8>,
+        block_header: Vec<u8>,
+    ) -> Self {
+        Self {
+            did_doc,
+            did_doc_meta,
+            chunk_file,
+            provisional_index_file,
+            core_index_file,
+            transaction,
+            merkle_block,
+            block_header,
+        }
+    }
+}
+
+/// Struct for Trustchain Verifier implementation via the ION DID method.
 pub struct IONVerifier<T>
```

Comment:
Consider whether we should include an `Option<URL>` field for an optional endpoint to receive a verification bundle. The `new()` method can include parsing. The `fetch_bundle()` method would then use field instead of receiving as an argument.

Reply:
As `fetch_bundles()` is already `pub` we can simply call this first (with a given endpoint) to cache before calling `IONVerifier.verify`.